junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,447 @@
+
+        val newClazz =
+          gen_mkClassDef_compat(Modifiers(), clazz.name, Nil, newImpl)
+        val clazzSym = {
+          val clazzSym = clazz.symbol
+          val newClazzInfo = {
+            val newParentsInfo = clazzSym.info.parents :::
+              typeOf[ScalaJSJUnitTest] :: Nil
+            val decls = clazzSym.info.decls
+            decls.enter(invokeJUnitMethodDef.symbol)
+            ClassInfoType(newParentsInfo, decls, clazzSym.info.typeSymbol)
+          }
+          clazzSym.setInfo(newClazzInfo)
+          clazzSym
+        }
+        newClazz.setSymbol(clazzSym)
+
+        newClazz
+      }
+
+      def mkHookClass(clazz: ClassDef, mlDefOption: Option[ClassDef]): Tree = {
+        val clazzSym = clazz.symbol.cloneSymbol
+        val getJUnitMetadataDef = mkGetJUnitMetadataDef(clazz, mlDefOption)
+        val newInstanceDef = mkNewInstanceDef(clazz, mlDefOption)
+        val invokeJUnitMethodDef = {
+          val annotatedMethods = mlDefOption.fold[List[DefDef]](Nil)(jUnitAnnotatedMethods)
+          val localMethodSymbol = (annotatedMethods ::: getMethods(clazz)).head.symbol
+          localMethodSymbol.setInfo(localMethodSymbol.info.atOwner(clazzSym))
+          mkInvokeJUnitMethodDef(annotatedMethods, localMethodSymbol)
+        }
+
+        val newBody = getJUnitMetadataDef :: newInstanceDef :: invokeJUnitMethodDef :: Nil
+
+        val newParents = TypeTree(typeOf[java.lang.Object]) ::
+          TypeTree(typeOf[JUnitTestMetadata]) :: Nil
+
+        val newImpl = clazz.impl.copy(parents = newParents, body = newBody)
+
+        val newClazz = {
+          val newClassName = newTypeName(clazz.name.toString + "$scalajs$junit$hook")
+          val newClazz = gen_mkClassDef_compat(Modifiers(Flags.MODULE),
+            newClassName, Nil, newImpl)
+          clazzSym.flags += Flags.MODULE
+          clazzSym.withoutAnnotations
+          clazzSym.setName(newClassName)
+          val newClazzInfo = {
+            val newParentsInfo = typeOf[java.lang.Object] ::
+              typeOf[JUnitTestMetadata] :: Nil
+            val decls = clazzSym.info.decls
+            decls.enter(getJUnitMetadataDef.symbol)
+            decls.enter(newInstanceDef.symbol)
+            decls.enter(invokeJUnitMethodDef.symbol)
+            ClassInfoType(newParentsInfo, decls, clazzSym.info.typeSymbol)
+          }
+          clazzSym.setInfo(newClazzInfo)
+          scalaJSPlugin.PrepInteropComponent.registerModuleExports(clazzSym)
+
+          newClazz.setSymbol(clazzSym)
+          newClazz
+        }
+        typer.typedClassDef(newClazz)
+      }
+
+      def jUnitAnnotatedMethods(clDef: ClassDef): List[DefDef] = {
+        clDef.impl.body.collect {
+          case ddef: DefDef if hasJUnitMethodAnnotation(ddef) => ddef
+        }
+      }
+
+      def getMethods(clazzDef: ClassDef): List[DefDef] =
+        clazzDef.impl.body.collect { case dDef: DefDef => dDef }
+
+      def mkInvokeJUnitMethodDef(methods: List[DefDef],
```

Comment:
It would be good to document most methods with a Scala source code representation of the trees they generate.